warl_proposal/csr_template_draft1.yaml

```diff
@@ -0,0 +1,130 @@
+    msb: <hex>                          # msb index of the csr. max: 31, min:0
+    lsb: <hex>                          # lsb index of the csr. min: 31, min:0
+    implemented: <boolean>              # indicates if the user has implemented this csr or not. When False, all fields 
+                                        # below this will be trimmed.
+    reset-val: <hex>                    # reset value of the subfield.
+    type:                               # indicates the type of the field. Can be one of the following only:
+      RO: <hex>                         # indicates the field is readonly with this value. This value == reset-value.
+      WARL:                             # indicates this field is WARL type. Refer to warl_proposal for more info.
+        dependency_variables: [list]    
+        behavior:
+          - func1:
+              dependency_values: [list] 
+              warl-func:
+                mode: <string>          
+                legal: [list]           
+      WLRL: [list of ranges]            # indicates this field is WLRL and the list of legal values as defined by the spec
+  rv64:                                 # this node and its subsequent fields can exist if [M/S/U]XL value can be 2
+    exists: <boolean>                   # indicates if this field exists in rv64 mode or not                          
+  rv128:                                # this node and its subsequent fields can exist if [M/S/U]XL value can be 2
+    exists: <boolean>                   # indicates if this field exists in rv64 mode or not                          
+##############################################################################################################
+
+############################################ Examples ############################################
+# example of csr with subfields
+mtvec:                                
+  description: Machine trap vector base address
+  address: 0x305         
+  priv_mode: M
+  reset-val: 0x00000001 
+  rv32:                                 
+    exists: True                        
+    subfields:                          
+      - mode
+      - base
+      - wpri                            # there is not wpri field in the mtvec 
+    mode:                   
+      description: trap handling mode
+      msb: 1
+      lsb: 0
+      implemented: True 
+      reset-val: 0x1
+      type:                             
+        WARL:
+          dependency_variables: []
+          behavior:
+            - func1: 
+                dependency_values: []
+                warl_func:
+                  mode: "Unchanged"
+                  legal: [[0x3,0x0,0x0,0x1]]
+    base:
+      description: vector base address
+      msb: 31
+      lsb: 2
+      implemented: True 
+      reset-val: 0x0
+      type:                             
+        WARL: 
+          dependency_variables: [mtvec::mode]
```

Comment:
I feel like this format is bit complicated. We could define some simple constraint syntax to describe this
mode == 0 -> base inside [0x20000000, 0x20004000]
mode == 1 -> base inside [0x00000000, 0x20000000], base[5:0] == 0

I feel this is more compact and readable. It might make the parsing script a bit more complex, but I feel it's more flexible and friendly for the user. We can limit the syntax support to reduce the parsing complexity. It will be even better to define the syntax to be compatible with SV constraint(a very small subuset) so that I can be directly used in verification.

Reply:
The thing is your proposal is specific to just mtvec. This would end up creating a specific parser for mtvec. Might not be scalable.

I would prefer a more generic template. We should be able to account of dependencies on other csrs and their subfields when defining a WARL. 

Let me see if I can come-up with something different with the same generic-ness

Reply:
I don't think this is specific to MTVEC, it's a constraint syntax to describe the relationship between different fields.  "mode", "base" is just a valid field name under the CSR. It could be different name for other CSR.

Reply:
operator: == , inside (could be others, like >, <, >=, <= etc)
-> imply
operand: mode, base (field name in the CSR, checker should validate this)
expression: 0 (fixed value), [a,b] list of value, [a:b] range

Reply:
A few benefits of doing this:
- More readable.
- More flexible than fixed field representation
- Simply the YAML format
- Make enhancement easier. Just add support in the parser in order to support new syntax, no need to touch the YAML format. 

Cons:
- Adding complexity to the parser.

Reply:
I agree its too complicated right now.. let me spend some time on this and get back to you.

Reply:
So I sort of extended your suggestion and following are a few examples of what I have in mind. I have replaced the `==` in your proposal so that the lhs and rhs of the `->` can be evaluated in same fashion.

Example 1: legal values as a range. I am using `&` to denote merging of the two definitions to obtain a single legal value.
```
constraint:
  - [mtvec::mode] inside [0] -> base[29:6] inside [0x200000 : 0x400000] & base [5:0] inside [0x0]
```

Example 2: legal values as a list of distinct values
```
constraint:
  - [mtvec :: mode] inside [1] -> base[29:0] inside [0x1000, 0x2000, 0x3000]
```

Example 3: legal values as a bitmask. `bitmask` is followed by a list of exactly 2 fields: `[mask,fixedval]`
```
constraint:
  - [mtvec::mode] inside [2,3] -> base[29:0] bitmask [0x3FFFFF00, 0x000000AB] 
```

Example 4: supporting multiple styles within `inside`. In this example base can be either `0x10000000` or any value within `[0x0000:0xFFFF]`. Notice the `|` operator
```
constraint:
  - [mtvec::mode] inside [0] -> base[29:0] inside  [0x10000000, 0x0000 : 0xFFFF]
```
Example 5: using multiple dependent variables:
```
constraint:
  -[mtvec::mode] inside [0:2] && [mstatus::mpp] inside [3] ->  base[29:0] inside  [0x10000000]
```

What do you think this format looks like? Would it be tough to parse these, if so what should we restrict?

Reply:
I think this format looks good, example 5 looks more complicated than the other one, we could limit multiple expressions like this if there's no use case. It's not supported by the original format anyway. We could use some special syntax to lower the parser complexity. For example, we can always use {...} for valuesand <...> for field name.

Reply:
great. Let me update the doc accordingly and formalize on this. Will have to get feedback from other folks as well

Reply:
Thanks,  just clarify the feature priority from my side, in case you want to make some trade off on the complexity:
1) Range 
    - Mandatory
2) WARL behavior, predict the read value for different write. 
    - bitmask support is mandatory, other prediction model is optional
3) constraint
    - Nice to have
4) Shadow
    - Mandatory

Reply:
Sure. Updated with new draft. Kindly let me know your thoughts on it.
Resolving this for now.

Reply:
I have a question on bitmask. How would define the wr_illegal nodes for the following 2 cases:
```
WARL:
  dependency_fields:
  legal:
    - "base[29:0] bitmask [0xFFF00, 0x000CD]
```

```
WARL:
  dependency_fields:
  legal:
    - "base[29:6] in [0x00000, 0x00FFF] & base[5:0] bitmask [0x01, 0x30]
```
bitmask shouldn't have any set of illegal values right? So i think for the first one `wr_illegal` node should be empty. But what about the second one?

I believe it would be convenient to restrict using bitmask in parts of the field (like the second example). either the entire field is bitmask or its defined using ranges for each sub-part

Would that be too restrictive?

Reply:
My understanding is that the bitmask is used to mask the read only bits within a WARL field. For example, if we want to tie off the lower 8 bits of the BASE.  It should be like this

BASE:
  bitmask : 0xFFF00

The value of the masked bits are determined by the reset value.  The mask should be applied to all the legal constraints.  BTW, the 2nd example seems to be too complex to parse. If "&" is supported, what about other logical operations?

Reply:
If the bitmask behavior is more complicated, like masking different bits based on the mode, you will need some way to describe this dependency. To me, the intent of describing the complete WARL behavior is an ambitious goal for this format. It could either add complexity to the parser or the format itself. That's why I said care more about basic things like range, shadow, basic bitmask. My suggestion on the format is mainly to achieve your target with a more user friendly way. Basically moving complexity from the YAML structure to the parser. However, fully describing WARL in this format is a kind of non-goal for me, it is useful for other flow though.

Reply:
instead of looking at the reset-val for the masked bits we expect a fixedval in the bitmask ops. This way the parser takes up only the WARL node. The reset-val is simply checked if it lies within the legal values or not.

The '&' is just a replacement for the ',' you had in your initial format.
We will support only '&' and no other operators. It indicates a merge of ops that's all.

I understand your perspective on this, but since we are targetting a larger audience with a variety of uses we need to be slightly more generic and flexible in a few areas. I agree WARL is an ambitious goal here, and we have had multiple iterations on this as well. I feel the proposed format is much intuitive to write and scalable as well.

Reply:
Sounds good to me.